Add proposal for Prometheus metrics coverage

terrytangyuan opened the pull request:
This provides a detailed outline of the Prometheus metrics we plan to coverage in common operator. Related issue: https://github.com/kubeflow/common/issues/22.

Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

terrytangyuan commented:
/cc @ywskycn @Jeffwan @gaocegege @richardsliu @johnugeorge @merlintang @jian-he @carmark

gaocegege commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,58 @@
+
+The following metrics related to the lifecycle of pods will be reported:
+
+* The total number of created pods
```
I think we can list the name of the metrics here.

Jeffwan replied:
Yeah. That would be great. We can use them directly in implementation.

terrytangyuan replied:
Good suggestion. Added.

Jeffwan commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,58 @@
+* Network usage
+* I/O usage
+* Keep-Alive check
+* Is-leader check
```
the leader means training job leader?

terrytangyuan replied:
I renamed it and added description. Please let me know if it's still unclear.

Jeffwan commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,58 @@
+
+## Job Metrics
+
+The following metrics related to the lifecycle of jobs will be reported:
```
I think we also need duration for some job state transformation. Like created -> started, started to finished, etc.

ywskycn replied:
+1, duration among different states is quite useful when debugging in production.

terrytangyuan replied:
Thanks. Added.

merlintang commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,58 @@
+* The total number of restarted pods
+* The total number of deleted pods
+* The total number of failed pods
+
```
How about the metric of job itself ? like the job start time, and end time?

ywskycn replied:
this can be covered by @Jeffwan 's suggestions.

terrytangyuan commented:
Thanks everyone for the comments! I've converted the lists to tables which include the metric name, type, and description. I also added a few additional metrics as suggested. Hopefully it's much clearer now. Please take another look.

johnugeorge commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,82 @@
+| container_network_transmit_bytes_total | Counter | Cumulative count of bytes transmitted |
+| container_fs_usage_bytes | Gauge | Number of bytes that are consumed by the container on this filesystem |
+| up | Gauge | Keep-Alive check (maintained by Prometheus on its own with its `up` metric detailed in the documentation [here](https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series))) |
+| common_operator_is_leader | Gauge | Whether this client is the leader of this common operator client set |
```
do you mean, the elected leader when redundant operator pods created ? How do you define this from Jobcontroller library perspective?

terrytangyuan replied:
Yes, this might be tricky as it involves moving the leader election related code to kubeflow/common. We can certainly do it at other operators if it doesn't fit in JobController.

johnugeorge replied:
Yes. Currently, this is not part of kubeflow/common. we could keep this out of scope for current time being.

terrytangyuan replied:
Sounds good. I removed it from the table.

Jeffwan commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,81 @@
+| up | Gauge | Keep-Alive check (maintained by Prometheus on its own with its `up` metric detailed in the documentation [here](https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series))) |
+
+Note that some of the above metrics are derived from [cAdvisor](https://github.com/google/cadvisor) kubelet
+integration which reports to Prometheus through our prometheus-operator installation.
```
Want to make sure the scope. This is outside operator. By default cadvisor expose the metrics and user can use these by their own.

terrytangyuan replied:
Yes, but I think it's good to document this here so we know that we don't need to report these metrics by ourselves.

Jeffwan commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,81 @@
+
+| Metric Name | Metric Type | Description |
+| ----------- | ------------| ----------- |
+| from_created_to_completed_job_duration_seconds_total | Counter | The duration between job created and job completed in seconds |
```
Minor: I am thinking if we should change to `job_duration_from_created_to_complated_seconds_total`.  Another thing is seems it would be good to use `complete` `deleted` as labels, but duration requires two and it would be a little bit hard to query. I think adding labels into metrics to distinguish them makes sense.

terrytangyuan replied:
I am following the naming practice outlined here: https://prometheus.io/docs/practices/naming/. I prefer the current naming without label as it's more intuitive but we can certainly revisit/revise later.

Jeffwan commented:
Beside above minor comments, it looks good to me. Wait to see if someone else has the feedback

terrytangyuan commented:
/assign @gaocegege @johnugeorge

yeya24 commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,81 @@
+| restarted_jobs_total | Counter | The total number of restarted jobs |
+| pending_jobs_total | Counter | The total number of pending jobs |
+| failed_jobs_total | Counter | The total number of failed jobs |
+| running_jobs_total | Counter | The total number of running jobs |
```
I am curious about this one. Why `running job` is a counter? Counter is usually used with `rate`, but as for `running job`, we want to know the current jobs in the running status. So I think it is more appropriate to use `Gauge` type, WDYT?

yeya24 replied:
The same for `pending jobs`

terrytangyuan replied:
Good catch. Thanks. Fixed!

yeya24 commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,81 @@
+| from_created_to_completed_job_duration_seconds_total | Counter | The duration between job created and job completed in seconds |
+| from_completed_to_deleted_job_duration_seconds_total | Counter | The duration between job completed and job deleted in seconds |
+| from_failed_to_restarted_job_duration_seconds_total | Counter | The duration between job failed and job restarted in seconds |
+| from_pending_to_running_job_duration_seconds_total | Counter | The duration between job pending and job running in seconds |
```
Why these metrics are `Counter`? Maybe for duration metrics, it is better to use `histogram`?

terrytangyuan replied:
Great suggestion. Updated.

terrytangyuan commented:
@yeya24 Thanks! Great suggestions. I have updated the metric types in the doc.

PTAL @gaocegege @johnugeorge

yeya24 commented on docs/prometheus-metrics.md:
```diff
@@ -0,0 +1,81 @@
+| completed_jobs_total | Counter | The total number of completed jobs |
+| restarted_jobs_total | Counter | The total number of restarted jobs |
+| pending_jobs_total | Gauge | The total number of pending jobs |
+| failed_jobs_total | Counter | The total number of failed jobs |
```
@terrytangyuan  Forgot to mention this one. Do you think it is more appropriate to make this `Gauge` as well?  Do you want to represent the history failures or the current failed jobs?

Can we list the metrics label in this doc as well? This is important and useful, too. Like we can combine `pending jobs` `running jobs` and `failed jobs` into one metric job_status{status="pending/failed/running"}, WDYT?

terrytangyuan replied:
Let's keep it as it is for now so that the metrics are consistent for metrics with past tense v.s. metrics with present continuous tense. Currently there are no labels yet as it's hard to differentiate metrics with two different tenses and choose different metric types for those metrics.

Jeffwan commented:
/lgtm

terrytangyuan commented:
/approve
